amworkflow/gcode/gcode.py

```diff
@@ -35,7 +39,263 @@ def create(self, in_file: Path, out_gcode: Path) -> None:
+            file.write(str(self.MaxLiquid2LevelRange1To7) + "\n")
+            file.write("\n"*67)   # Values from 34 to 100 are not existant, therefore empty lines
+
+        ### Write body - voxel to hex
+        # Check if there are more voxels along x than nozzles
+        if voxel_3d_array.shape[0] > self.NozzleNum:
+            raise ValueError(f"{voxel_3d_array.shape[0]} x-voxels to only {self.NozzleNum} nozzles, rescale model to fit within printer width of {self.PrinterY}m")
+        # Check if 4 bit to hex encoding possible
+        if self.NozzleNum % 4 != 0:
+            raise ValueError("Number of nozzles not a multiple of 4, 4bit to hex encoding not possible")
+        
+        # Prepare array
+        printerbed_3d_array = np.zeros((self.NozzleNum, self.LinesNum, sorted_z.shape[0]))
+
+        # Sort into temporary array
+        for ii in range(voxel_3d_array.shape[2]):
+            # Position smaller prints into larger printer bed, roughly in the middle
+            A_large = printerbed_3d_array[:,:,ii]
+            A_small = voxel_3d_array[:,:,ii]
+            start_row = (A_large.shape[0] - A_small.shape[0]) // 2
+            start_col = (A_large.shape[1] - A_small.shape[1]) // 2
+
+            A_large[start_row:start_row + A_small.shape[0], start_col:start_col + A_small.shape[1]] = A_small
+            printerbed_3d_array[:,:,ii] = A_large
+
+        # Print the actual dsmn-file
+        for layers in range(printerbed_3d_array.shape[2]):
+            printer_hex = []
+            temp_array = np.transpose(printerbed_3d_array[:,:,layers])
+
+            for row in temp_array:
+                # Partitioning of rows in parts of length 4
+                chunks = [row[i:i+4] for i in range(0, len(row), 4)]
+
+                # Convert every chunk into binary number and then into hex number
+                hex_values = [hex(int(''.join(map(str, chunk.astype(int))), 2))[2:].upper() for chunk in chunks]
+
+                # Combine hex numbers
+                printer_hex.append(hex_values)
+
+            with open(out_dsmn, "a") as file:
+                for hex_values in printer_hex:
+                    file.write("".join(hex_values) + "\n")
+                if self.DebugMode:
+                    file.write("-" * (int(self.NozzleNum/4)) + "\n") # line separator between layers
+
+        # # Visualize voxels with pyvista
+        # def visualize_voxels(voxel_centers):
+        
+        #     # Create a point cloud representation
+        #     point_cloud = pv.PolyData(voxel_centers) #pv.PointCloud(voxel_polydata)
+
+        #     # Plot the point cloud using pyvista plotter
+        #     point_cloud.plot(eye_dome_lighting=True)
+        # visualize_voxels(voxel_array)
+
+        ### Print some data for the user to file
+        orig_stdout = sys.stdout
+        f = open(log_file_path, "w")
+        sys.stdout = f
+        print("-- Printing and slicing statistics --")
```

Comment:
change prints to logger output -> self.logger.info('blabla') or debug or error dependening on information